lib/FrontendTool/FrontendTool.cpp

```diff
@@ -871,6 +937,217 @@ static bool performCompile(CompilerInstance &Instance,
+    // to run the ownership evaluator.
+    if (runSILOwnershipEliminatorPass(*SM))
+      return true;
+  }
+
+  linkAllIfNeeded(Invocation, SM);
+
+  if (Invocation.getSILOptions().MergePartialModules)
+    SM->linkAllFromCurrentModule();
+  return false;
+}
+
+static SerializationOptions
+computeSerializationOptions(const CompilerInvocation &Invocation,
+                            bool moduleIsPublic) {
+  const FrontendOptions &opts = Invocation.getFrontendOptions();
+
+  SerializationOptions serializationOpts;
+  serializationOpts.OutputPath = opts.ModuleOutputPath.c_str();
+  serializationOpts.DocOutputPath = opts.ModuleDocOutputPath.c_str();
+  serializationOpts.GroupInfoPath = opts.GroupInfoPath.c_str();
+  if (opts.SerializeBridgingHeader)
+    serializationOpts.ImportedHeader = opts.ImplicitObjCHeaderPath;
+  serializationOpts.ModuleLinkName = opts.ModuleLinkName;
+  serializationOpts.ExtraClangOptions =
+      Invocation.getClangImporterOptions().ExtraArgs;
+  serializationOpts.EnableNestedTypeLookupTable =
+      opts.EnableSerializationNestedTypeLookupTable;
+  if (!Invocation.getIRGenOptions().ForceLoadSymbolName.empty())
+    serializationOpts.AutolinkForceLoad = true;
+
+  // Options contain information about the developer's computer,
+  // so only serialize them if the module isn't going to be shipped to
+  // the public.
+  serializationOpts.SerializeOptionsForDebugging =
+      !moduleIsPublic || opts.AlwaysSerializeDebuggingOptions;
+
+  return serializationOpts;
+}
+
+/// Perform SIL optimization passes if optimizations haven't been disabled.
+/// These may change across compiler versions.
+static void performSILOptimizations(CompilerInvocation &Invocation,
+                                    SILModule *SM) {
+  SharedTimer timer("SIL optimization");
+  if (Invocation.getFrontendOptions().RequestedAction ==
+          FrontendOptions::ActionType::MergeModules ||
+      !Invocation.getSILOptions().shouldOptimize()) {
+    runSILPassesForOnone(*SM);
+    return;
+  }
+  runSILOptPreparePasses(*SM);
+
+  StringRef CustomPipelinePath =
+      Invocation.getSILOptions().ExternalPassPipelineFilename;
+  if (!CustomPipelinePath.empty()) {
+    runSILOptimizationPassesWithFileSpecification(*SM, CustomPipelinePath);
+  } else {
+    runSILOptimizationPasses(*SM);
+  }
+}
+
+/// Get the main source file's private discriminator and attach it to
+/// the compile unit's flags.
+static void setPrivateDiscriminatorIfNeeded(IRGenOptions &IRGenOpts,
+                                            ModuleOrSourceFile MSF) {
+  if (IRGenOpts.DebugInfoKind == IRGenDebugInfoKind::None ||
+      !MSF.is<SourceFile *>())
+    return;
+  Identifier PD = MSF.get<SourceFile *>()->getPrivateDiscriminator();
+  if (!PD.empty())
+    IRGenOpts.DWARFDebugFlags += (" -private-discriminator " + PD.str()).str();
+}
+
+static bool serializeSIBIfNeeded(FrontendOptions &opts, SILModule *SM,
```

Comment:
It's "needed" at both call sites, so you can probably turn that early exit into an assertion.

Reply:
Could you please clarify? I think it's doing the same as the old code:
`// Get the main source file's private discriminator and attach it to
  // the compile unit's flags.
  if (IRGenOpts.DebugInfoKind != IRGenDebugInfoKind::None &&
      MSF.is<SourceFile*>()) {
    Identifier PD = MSF.get<SourceFile*>()->getPrivateDiscriminator();
    if (!PD.empty())
      IRGenOpts.DWARFDebugFlags += (" -private-discriminator "+PD.str()).str();
  }`

Reply:
This is about `serializeSIBIfNeeded` not the private discriminator one.

Reply:
Sorry. Fixed, and removed the no-longer-correct "IfNeeded".